agent/testflinger_agent/runner.py

```diff
@@ -0,0 +1,128 @@
+# Copyright (C) 2024 Canonical
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>
+
+import logging
+import os
+import fcntl
+import sys
+import shlex
+import signal
+import subprocess
+import threading
+import time
+
+from typing import Callable, Optional, List
+
+logger = logging.getLogger(__name__)
+
+
+class CommandRunner:
+    def __init__(self, cwd: Optional[str], env: Optional[dict]):
+        self.output_handlers: List[Callable] = []
+        self.stop_condition_checkers: List[Callable] = []
+        self.process: Optional[subprocess.Popen] = None
+        self.cwd = cwd
+        self.env = os.environ.copy()
+        if env:
+            self.env.update(
+                {k: str(v) for k, v in env.items() if isinstance(v, str)}
+            )
+
+    def register_output_handler(self, handler: Callable[[str], None]):
+        self.output_handlers.append(handler)
+
+    def post_output(self, data: str):
+        for handler in self.output_handlers:
+            handler(data)
+
+    def register_stop_condition_checker(
+        self, checker: Callable[[], Optional[str]]
+    ):
+        self.stop_condition_checkers.append(checker)
```

Comment:
Same point here, what do you think of defining a type alias for stop condition checker?

Reply:
Added `StopConditionType`